File: php/test/SslEntityTest.php
<?php
declare(strict_types=1);

// Ssl entity test

require_once __DIR__ . '/../whoisdomainmonitoring_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class SslEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = WhoisDomainMonitoringSDK::test(null, null);
        $ent = $testsdk->Ssl(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "ssl" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = WhoisDomainMonitoringSDK::test($seed, null);
        $seen = iterator_to_array($base->Ssl(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = WhoisDomainMonitoringConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = WhoisDomainMonitoringSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Ssl(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = ssl_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "ssl." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $ssl_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.ssl")));
        $ssl_ref01_data = null;
        if (count($ssl_ref01_data_raw) > 0) {
            $ssl_ref01_data = Helpers::to_map($ssl_ref01_data_raw[0][1]);
        }

        // LIST
        $ssl_ref01_ent = $client->Ssl(null);
        $ssl_ref01_match = [];

        $ssl_ref01_list_result = $ssl_ref01_ent->list($ssl_ref01_match, null);
        $this->assertIsArray($ssl_ref01_list_result);

    }
}

function ssl_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/ssl/SslTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = WhoisDomainMonitoringSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["ssl01", "ssl02", "ssl03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID" => $idmap,
        "WHOIS_DOMAIN_MONITORING_TEST_LIVE" => "FALSE",
        "WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN" => "FALSE",
        "WHOIS_DOMAIN_MONITORING_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["WHOIS_DOMAIN_MONITORING_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new WhoisDomainMonitoringSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: rb/test/ssl_entity_test.rb
# Ssl entity test

require "minitest/autorun"
require "json"
require_relative "../WhoisDomainMonitoring_sdk"
require_relative "runner"

class SslEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WhoisDomainMonitoringSDK.test(nil, nil)
    ent = testsdk.Ssl(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "ssl" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = WhoisDomainMonitoringSDK.test(seed, nil)
    seen = base.Ssl(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = WhoisDomainMonitoringConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = WhoisDomainMonitoringSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Ssl(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = ssl_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "ssl." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    ssl_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.ssl")))
    ssl_ref01_data = nil
    if ssl_ref01_data_raw.length > 0
      ssl_ref01_data = Helpers.to_map(ssl_ref01_data_raw[0][1])
    end

    # LIST
    ssl_ref01_ent = client.Ssl(nil)
    ssl_ref01_match = {}

    ssl_ref01_list_result = ssl_ref01_ent.list(ssl_ref01_match, nil)
    assert ssl_ref01_list_result.is_a?(Array)

  end
end

def ssl_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "ssl", "SslTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WhoisDomainMonitoringSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["ssl01", "ssl02", "ssl03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID" => idmap,
    "WHOIS_DOMAIN_MONITORING_TEST_LIVE" => "FALSE",
    "WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN" => "FALSE",
    "WHOIS_DOMAIN_MONITORING_APIKEY" => "",
  })

  idmap_resolved = Helpers.to_map(
    env["WHOIS_DOMAIN_MONITORING_TEST_SSL_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
        "apikey" => env["WHOIS_DOMAIN_MONITORING_APIKEY"],
      },
      extra || {},
    ])
    client = WhoisDomainMonitoringSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
